subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/JavaToolchainRepository.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+
+import java.net.URI;
+import java.util.Optional;
+
+/**
+ * Interface that needs to be implemented by Toolchain SPI plugins, in order to
+ * extend Gradle with their spec-to-URI logic.
+ *
+ * //TODO (#21082): more/better docs
+ *
+ * @since 7.6
+ */
+@Incubating
+public interface JavaToolchainRepository extends BuildService<BuildServiceParameters.None> {
+
+    /**
+     * Returns the URI from which a Java Toolchain matching the provided specification can be downloaded.
+     * The URI must point to either a ZIP or a TAR archive file.
+     *
+     * Returns an empty Optional if and only if the provided specification can't be matched.
+     */
+    Optional<URI> toUri(JavaToolchainSpec spec);
```

Comment:
What is the general opinion 'round here about `Optional`s? I rarely see these around the codebase, and `@Nullable` return types seem much more common. 

I've definitely got mixed feeling about them, as I am sure many others do, but I'm mostly concerned about consistency here with the rest of the Gradle API

Reply:
I would like to up this comment. Regardless of my personal preferences, I believe it is better to have a consistent API in relation to nullability. Same as @jvandort, I don't know whether there is an agreed on direction that new APIs should use `Optional`s.

Just as a side note, `Optional`s in Kotlin would create a bit more clutter compared to the properly annotated nullability.

Reply:
Good point, will consider it.

Reply:
I understand the question about using `Optional` but I really do not want to introduce an API where the contract is "return `null` if you cannot satisfy the toolchain spec".
A potentially more idiomatic approach would be to use `Provider` here.

Reply:
`Provider` is a bad idea, it carries other meaning.

Note that we added `Optional` API for users in 7.5: See `ConfigurationVariant`